### PR TITLE
Moving New-Guid to compiled cmdlet

### PR DESCRIPTION
**Pull request by powercode**
Fixing #2733 
Simplest implementation possible

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Utility/NewGuid.Tests.ps1
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/NewGuid.Tests.ps1]: should be?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: We indent with spaces, not tabs.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: The cmdlet suggests that it creates a new object every time it's called, but with this parameter, nothing is actually created.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: This code feels like it doesn't belong in a cmdlet with the verb `New`, and I don't think we want a cmdlet `Convert-ToGuid`.

**Review by lzybkr — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: How I wish that tab/spaces was a per project setting in VS...
DICE has tabs, and switching back and forth is a mess...

Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: fixed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: Well, you have noting before, and after you have a guid. Admittedly a static one that will never change, but it is friendlier to type new-guid -e than [guid]::empty, especially with Scandinavian keyboards, where '[' is on AltGr+8.

But it's your call. Remove?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: removed

File: src/Microsoft.PowerShell.Commands.Utility/map.json
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/map.json]: You don't need to update `map.json` for new files.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: (Copy from Issue) Some applications use GUID as a unique identifier. In this case, the identifier is a string. So really, we can get a stream of strings (ex. from logs), that we want to convert to  [guid] . And yes this is more for  ConvertTo-Guid  but I looked through  New-*  cmdlets and many of them do a similar operation (string -> object). Most likely the new  ConvertTo-Guid  cmdlet is redundant and perhaps it would be better to make the  GUID  parameter.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: It seems this parameter is redundant.
The only cmdlet that uses `Count` parameter is `Get-Random` and only to determines how many objects are returned from input pipeline.
Usually we do so:
```powershell 
0..9 | New-Guid
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: Most likely you had in mind:
`if (String.Equals(ParameterSetName, "Empty", StringComparison.OrdinalIgnoreCase))
or
`if (ParameterSetName == "Empty")`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: Well, to me, that looks like the input in some way should be related to the created guids.

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: Yes, the sample is incorrect.
I mean:
```powershell
0..10 | % { Get-Random }
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: Using `if (Empty)` here will also handle a rare usage of the cmdlet: `New-Guid -Empty:$false` -- it will write out a new GUID in that case.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: This comment probably should be `Return one or more guids`.

**Comment by lzybkr** [comment]
Don't merge yet - I've marked this for committee review.

I personally wouldn't mind removing this command, or if we do, it should work just like the existing function, nothing more, but we should decide that as a committee.

**Comment by iSazonov** [comment]
I analyzed using `NewGuid` by @daxian-dbw [method ](https://gist.github.com/daxian-dbw/07847ff5ef6da2c100e8b22bd94f0feb)
Found **133** *.ps1 scripts on GitHub (I did not search *.psm1).
Thus it is very rarely used cmdlet.

I manually inspected these scripts.
I did not find that this cmdlet was used in loops.
I did not find using `Empty` guids.

A common usage (in descending order of popularity):
```powershell
$GUID = (New-Guid).Guid

[string]$Name = (New-Guid).ToString()

$id = New-Guid
$name = "Solution_$id"

$wpconfig = $wpconfig.Replace("eisauthkey",$(New-Guid))

function New-Guid {
    [System.Guid]::NewGuid().ToString("d").Substring(0, 4).Replace("-", "")
}

$guid = (New-Guid).GUID.Replace("-","_")
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (JamesWTruher) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: >EndProcessing [](start = 32, length = 13)

this implementation subtly changes the current implementation (the function uses Begin rather than End as is here) - I don't believe it to be a big problem, but thought I should add the note

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee we discussed this and for discovery reasons as well as shipping this previously, we should have new-guid even though it's a simple operation for users familiar with the guid class

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewGuidCommand.cs]: This is intentional, see this discussion: https://github.com/PowerShell/PowerShell/pull/2786/files/48a6fb2e05c6d4954f94d9f34139ff338aae41a2#r89893087

**Review by lzybkr — APPROVED** [review]

**Pull request merged**
